### PR TITLE
Add support for DirectCounter in table add/update

### PR DESCRIPTION
**Pull request by tatry**

### CODE REVIEW

File: lib/psabpf_table.c [hunk 1]
```diff
@@ -55,6 +55,89 @@ void psabpf_table_entry_ctx_free(psabpf_table_entry_ctx_t *ctx)
...
+    }
+    ctx->n_direct_counters = 0;
+}
+
+enum direct_object_command {
+    DIRECT_OBJECT_COUNT,
+    DIRECT_OBJECT_INIT_CTX,
+};
+
+static int execute_direct_objects_command(psabpf_table_entry_ctx_t *ctx, enum direct_object_command command)
```
Comment (kmateuszssak) [inline lib/psabpf_table.c:74]: I'm not sure if `command` terminology is ok here. I would call this operation adding... I don't see for now good choice...

File: lib/psabpf_table.c [hunk 2]
```diff
@@ -117,6 +200,24 @@ int psabpf_table_entry_ctx_tblname(psabpf_context_t *psabpf_ctx, psabpf_table_en
...
 
+    if (ctx->btf_metadata.btf == NULL || ctx->table.btf_type_id == 0) {
+        fprintf(stderr, "unable to handle direct objects; resetting them if exist\n");
+    } else {
+        ret = execute_direct_objects_command(ctx, DIRECT_OBJECT_COUNT);
+        if (ret == NO_ERROR) {
+            if (ctx->n_direct_counters > 0) {
+                ctx->direct_counters_ctx = malloc(ctx->n_direct_counters * sizeof(psabpf_direct_counter_context_t));
+                for (unsigned i = 0; i < ctx->n_direct_counters; i++)
+                    psabpf_direct_counter_ctx_init(&ctx->direct_counters_ctx[i]);
```
Comment (kmateuszssak) [inline lib/psabpf_table.c:211]: `psabpf_direct_counter_ctx_init` and `DIRECT_OBJECT_INIT_CTX` command confuse me

File: lib/psabpf_table.c [hunk 3]
```diff
@@ -959,96 +1068,60 @@ static int construct_buffer(char * buffer, size_t buffer_len,
     return return_code;
 }
 
-static bool member_is_counter(psabpf_table_entry_ctx_t *ctx, const struct btf_member *member)
+static int handle_direct_objects(const char *key, char *value,
```
Comment (kmateuszssak) [inline lib/psabpf_table.c:1071]: I suggest to create separated methods for counters and meters in `handle_direct_objects`

**Review by kmateuszssak — CHANGES_REQUESTED** [review]
As always only minor comments :)

File: CLI/table.c [hunk 4]
```diff
@@ -181,6 +210,22 @@ static int parse_action_data(int *argc, char ***argv,
                 ref_is_group_ref = true;
                 NEXT_ARGP_RET();
             }
+        } else {
```
[the remark this answers is not in the document]
Comment (kmateuszssak) [reply CLI/table.c:213]: Ok, there is some logic behind this.

File: CLI/table.c [hunk 5]
```diff
@@ -159,8 +160,36 @@ static int parse_table_key(int *argc, char ***argv, psabpf_table_entry_t *entry)
...
-                             psabpf_action_t *action, bool indirect_table)
+static int parse_direct_counter_entry(int *argc, char ***argv,
+                                      psabpf_table_entry_ctx_t *ctx, psabpf_table_entry_t *entry,
+                                      psabpf_direct_counter_context_t *dc, psabpf_counter_entry_t *counter)
+{
+    if (!is_keyword(**argv, "counter"))
+        return EINVAL;
+
+    NEXT_ARGP_RET();
+    const char *instance = **argv;
```
[the remark this answers is not in the document]
Comment (tatry) [reply CLI/table.c]: Done

File: lib/psabpf_table.c (hunk 1 above)
Comment (tatry) [reply lib/psabpf_table.c]: Refactored

File: lib/psabpf_table.c (hunk 2 above)
Comment (tatry) [reply lib/psabpf_table.c]: Moved to `init_direct_objects`

File: lib/psabpf_table.c (hunk 3 above)
Comment (tatry) [reply lib/psabpf_table.c]: Done

**Comment by kmateuszssak** [comment]
LGTM!

**Pull request merged**